### ISSUE TITLE
[Feature request] Support transactions

### ISSUE THREAD

**Issue by claustres** [feathersjs-ecosystem/feathers-mongodb#127]
Transactions are available in Mongo V4: https://www.mongodb.com/transactions & https://thecodebarbarian.com/a-node-js-perspective-on-mongodb-4-transactions.html.

A way to ease implementing it is to provide:
* before hook to start the transaction
* an after hook to commit the transaction
* an error hook to cancel the transaction

Because we also need to pass the transaction session into each DB call as well the module should also support passing it as a param and send it to Mongo operations.

**Comment by eddyystop** [comment]
Should this be in core or common hooks? Common hooks has already moved into DB-specific hooks with mongokeys https://feathers-plus.github.io/v1/feathers-hooks-common/#mongokeys

**Comment by eddyystop** [comment]
Maybe base it on if the hook is fundamentally core, or a nice to have plus.

**Comment by easonzhou** [comment]
Is there any progress on this request?

**Comment by barnacker** [comment]
I'm interested in this feature as well or a suggestion for implementation :)

**Comment by claustres** [comment]
It seems to me that some ways of implementing have been clearly exposed so what is missing is only someone with enough time to start implementing, PR are welcome as usual ;-)

**Comment by arlukin** [comment]
Has anybody done anything about this? I started to experiment a bit with transactions in feathers-mongodb. Hooks looked like the right way to do it for me. Only need transactions for some patch() calls in my rest-api.

This gist is my current draft. https://gist.github.com/arlukin/28f99a3cbcec4c907e76d15e5f3c3077
Not really something that can be added as a PR.

Three things I'm not really satisfied with.

* Had to add params.inTransaction to know if I was in a transaction. MongoSession has a inTransaction() function. But can't find a way to get the MongoClient or the MongoSession from the MongoCollection that is available in _get trough this.modal. 
http://mongodb.github.io/node-mongodb-native/3.5/api/ClientSession.html#inTransaction

* I don't yet fully understand of how mongodb sessions work with nodejs async/one-thread. If I should use one session per request, as I'm doing right now. Or if I can use one session on global-scope when I'm doing the MongoClient().connect(). 

* I don't know how to implement support for TransientTransactionError. See runTransactionWithRetry in https://docs.mongodb.com/manual/core/transactions-in-applications/#core-api

This article can also be nice to read to understand why I had to use findOneAndUpdate in _get()
https://www.mongodb.com/blog/post/how-to-select--for-update-inside-mongodb-transactions

**Comment by claustres** [comment]
Great you start working on this. 

Concerning the required access to the mongo client for transactions I would recommend to make it an options of the hook instead of reading it from the config. Indeed, you will find use cases where there are multiple clients connected to the same or different databases in the app. Like this you would be able to possibly select a different client for different services. Moreover, it is also consistent with the adapter behavior: when you create a service you simply pass the collection to the service, not the client or the DB so that your collections can come from different databases if required, it's transparent from the module point of view. Similarly, transaction options should be part as the hook options as we can have different use cases as well regarding consistency, this can become tricky with replica sets and sharding.

About sessions, the doc states that there is only one open transaction for a session https://docs.mongodb.com/manual/core/transactions/#transactions-and-sessions, meaning that if you want to allow multiple transactions in parallel you will need multiple sessions. Usually operations of different sessions could go in parallel without frequent blocking, but once again it can strongly depends on the use case and the DB setup. So I would say that we should allow for different sessions in parallel like you do now, maybe this can also be an option so that if we don't allow parallelism we have a single global session in this case instead of a session per context.

About the transient error this simply means that an error that cannot be recovered has been raised during the transaction, in that case you should simply raise an error as the error management will be an app specific logic.

Last, ensuring consistency between read/write operations in a transaction (https://www.mongodb.com/blog/post/how-to-select--for-update-inside-mongodb-transactions) is probably a specific use case and I am not sure it is a good idea to hard-code it in the get operation. Maybe it would be better to make it a separate option in `params`, indeed if your transaction only consists in eg patching specific fields on a set of documents you don't really need to read before, you might want the operation to simply fail if there is a write conflict.

I hope this helps.

**Comment by arlukin** [comment]
Sorry @claustres for not answering you back in may. Summer and a heap of todos got in the way.

Out first attempt to handle transactions was a bit naive. Things was a bit more complicated than I first thought. I have spent some time the last weeks to work on a couple of new hooks and using them in our application. It needs som more refactoring, optimization and documentation, but they seem to work with our use cases.

I followed you recommenation and added mongoClient to the start hook. It kind of made more sense when coding everything in a fork of feathers-mongodb.

I also removed the hard-coded get operation. The row lock is now handled by 3 hooks. lockData, loadData and patchData.

The thought about how to work with the hooks is probably best seen in the unit test "transactions.test.js".

Here is a compare between my fork and the main code. It's not ready for a pull request, but I hope this can help anyone.
https://github.com/feathersjs-ecosystem/feathers-mongodb/compare/master...arlukin:master
